Refactor the memtest module and add more memtests

BrianShTsoi opened the pull request:
Refactor by adding the `TestObserver` trait and `MemtestResult` type alias.
Also added ~~6~~ 7 new memtests adapted from [Memtest86+](https://github.com/memtest86plus/memtest86plus).

BrianShTsoi committed:
Implement TestObserver trait to replace TimeoutChecker in memtest functions

BrianShTsoi committed:
Add 6 memtests with Moving Inversion and Modulo-20 algorithms

BrianShTsoi committed:
Implement the MemtestResult type alias

BrianShTsoi committed:
Reformat memtest.rs

BrianShTsoi commented on src/memtest.rs:
```diff
@@ -407,36 +387,392 @@ pub fn test_checkerboard(
+/// This test uses the Modulo-20 algorithm implemented by [memtest86+](https://github.com/
+/// memtest86plus/memtest86plus),  which is designed to avoid effects of caching and buffering.
+///
+/// The test generates a random value, then write the value to every 20th memory location.
+/// Afterwards write the complement of the value to all other locations. Then verify that the values
+/// stored in every 20th location is unchanged.
+/// The procedure is repeated with offsets 0-19 to test all memory locations.
```
```suggestion
/// The test generates a random value, then write the value to every 20th memory location.
/// Afterwards write the complement of the value to all other locations one or more times (twice in
/// this case). Then verify that the values stored in every 20th location is unchanged.
///
/// The procedure is repeated with offsets 0-19 to test all memory locations.
```
Small correction to comments

BrianShTsoi commented on src/lib.rs:
```diff
@@ -315,7 +325,10 @@ impl MemtestReportList {
 
 impl MemtestReport {
-    fn new(test_kind: MemtestKind, outcome: Result<MemtestOutcome, MemtestError>) -> MemtestReport {
+    fn new(
+        test_kind: MemtestKind,
+        outcome: Result<MemtestOutcome, MemtestError<TimeoutError>>,
+    ) -> MemtestReport {
```
```suggestion
    fn new(
        test_kind: MemtestKind,
        outcome: MemtestResult<TimeoutChecker>,
    ) -> MemtestReport {
```

marti4d replied:
Good catch!

marti4d commented on src/lib.rs:
```diff
@@ -66,7 +66,7 @@ pub struct MemtestReportList {
 pub struct MemtestReport {
     pub test_kind: MemtestKind,
-    pub outcome: Result<MemtestOutcome, MemtestError>,
+    pub outcome: MemtestResult<TimeoutChecker>,
```
I'd say skip using your type alias here, because it requires you to leak an implementation detail even though that
detail has nothing to do with the type itself (you're just using it for its `Error` type).

Instead, I'd spell this out as `Result<MemtestOutcome, MemtestError<TimeoutError>>` and get rid of the `pub` on
`TimeoutChecker`.

marti4d commented on src/lib.rs:
```diff
@@ -91,7 +91,7 @@ struct MemLockGuard {
 /// A struct to ensure the test timeouts in a given duration
 #[derive(Debug)]
-struct TimeoutChecker {
+pub struct TimeoutChecker {
```
See above comment - This should not be needed.

marti4d commented on src/memtest.rs:
```diff
@@ -33,9 +34,15 @@ pub enum MemtestFailure {
 #[derive(Debug, Serialize, Deserialize)]
-pub enum MemtestError {
-    Timeout,
+pub enum MemtestError<E: Error> {
```
[This is a good read](https://stackoverflow.com/questions/49229332/should-trait-bounds-be-duplicated-in-struct-and-impl) to explain why you should almost-never put bounds on the generics in a `struct` or `enum` definition. 

TL;DR: Put them only on the `impl` blocks, and only the trait bounds that particular `impl` block needs to get the job done.

(If you continue reading my comments toward the end of this file on your `impl` for this type, that might also help make some sense).

marti4d commented on src/memtest.rs:
```diff
@@ -92,23 +105,23 @@ pub struct ParseMemtestKindError;
 
     for mem_ref in memory.iter_mut() {
-        timeout_checker.check()?;
+        observer.check().map_err(|e| MemtestError::Observer(e))?;
```
Should be `.map_err(MemtestError::Observer)`. Just a friendly reminder to make sure you run `cargo clippy` before putting your changes up for review :)

marti4d commented on src/memtest.rs:
```diff
@@ -407,36 +387,392 @@ pub fn test_checkerboard(
+    observer.init(expected_iter);
+
+    let mut pattern = usize_filled_from_byte(0x10);
+    for _ in 0..=(u8::try_from(NUM_RUNS - 1).unwrap()) {
```
Should be fine to just do `for _ in 0..NUM_RUMS`. I think all that casting nastiness is because you actually use the counter value in the other test that you copy-pasta'd from.

marti4d commented on src/memtest.rs:
```diff
@@ -407,36 +387,392 @@ pub fn test_checkerboard(
+/// to the description available at the [Memtest86+ Test Algorithm Section](https://github.com/
+/// memtest86plus/memtest86plus?tab=readme-ov-file#memtest86-test-algorithms)
+///
+/// This test runs the moving inversion algorithm with random fixed patterns.
```
nit: I would change this word to say "a random fixed pattern". The currently wording implies that it uses more than one random pattern.

marti4d commented on src/memtest.rs:
```diff
@@ -407,36 +387,392 @@ pub fn test_checkerboard(
+    memory: &mut [usize],
+    mut observer: O,
+) -> MemtestResult<O> {
+    const NUM_RUNS: usize = size_of::<usize>();
```
I think this is a sneaky logic error. `size_of()` returns the size of a type in bytes -- So `4` on 32-bit and `8` on 64-bit.

But you want to do a run for each **bit** position: `NUM_RUNS = 32` on 32-bit and `NUM_RUNS = 64` on 64-bit. You need to multiply by `8` :)

marti4d commented on src/memtest.rs:
```diff
@@ -407,36 +387,392 @@ pub fn test_checkerboard(
+        if let MemtestOutcome::Fail(f) = mov_inv_walking_pattern(memory, pattern, &mut observer)? {
+            return Ok(MemtestOutcome::Fail(f));
+        }
+        if let MemtestOutcome::Fail(f) = mov_inv_walking_pattern(memory, !(pattern), &mut observer)?
```
nit: unnecessary parans

marti4d commented on src/memtest.rs:
```diff
@@ -407,36 +387,392 @@ pub fn test_checkerboard(
+    (memory.len() > STEP)
+        .then_some(())
+        .context("Insufficient memory length for two-regions memtest")?;
+    let expected_iter = u64::try_from(memory.len())
```
This calculation must almost-certainly be incorrect? The answer must involve some kind of division, since on the last pass you only check every 20th memory address to see if it's correct.

BrianShTsoi replied:
I think the calculation was right but my comment was wrong. 
The test has 20 runs (steps), for each run, it writes to every 20th address, then writes to the other 19 addresses *twice*, then finally it verifies that values at every 20th address is unchanged. 
So each address is either read or write twice. Hence `memory.len() * STEP * 2`.

marti4d commented on src/memtest.rs:
```diff
@@ -510,23 +846,23 @@ impl fmt::Display for MemtestOutcome {
 }
 
-impl fmt::Display for MemtestError {
+impl<E: Error> fmt::Display for MemtestError<E> {
```
One of the benefits of getting rid of the bound on your `struct` is that you can have looser bounds on your `impl`s. For example, the function you wrote here will work for any type, `E`, as long as it implements the `fmt::Debug` trait (required by the `{:?}` formatting syntax). So, your impl bound here can simply be `impl<E: fmt::Debug> fmt::Display for MemtestError<E>`.

marti4d commented on src/memtest.rs:
```diff
@@ -510,23 +846,23 @@ impl fmt::Display for MemtestOutcome {
 }
 
-impl Error for MemtestError {
+impl<E: Error> Error for MemtestError<E> {
```
I'm pretty sure you *should* make it so the observer's error **is** the source of the `MemtestError`. You're going to need a `'static` lifetime in your trait bound for that, though, so: `impl<E: Error + 'static> Error for MemtestError<E>`.

marti4d commented on src/memtest.rs:
```diff
@@ -510,23 +846,23 @@ impl fmt::Display for MemtestOutcome {
 
-impl From<anyhow::Error> for MemtestError {
-    fn from(err: anyhow::Error) -> MemtestError {
+impl<E: Error> From<anyhow::Error> for MemtestError<E> {
```
You actually don't need any trait bound here: Since you're using the `Other(anyhow::Error)` variant and not the `Observer(E)` variant, it doesn't matter what `E` is. This is exactly one of the reasons why you don't declare trait bounds on `enum` -- You restrict what types your user can use even if they're not using the variant imposing the restriction.

marti4d reviewed (changes requested):
Hey Brian,

It's looking very good, overall. A few small comments here and there.

One thing I noticed is that I'm fairly certain the iteration count is wrong for one of the tests. Fixable, but it got me thinking: maybe we should be adding an assert to catch these logic errors. Basically, if a test passes, that means that it definitely did not bail out early, and so it ran all iterations.

I have an idea for how to write a unit test to catch this:

1.  Now that you have your observer trait, make a new type just for testing called something like `IterationCounter` and
    `impl TestObserver for &mut IterationCounter`.

2.  `IterationCounter` stores the `expected_iter` it gets passed in by `init`, and also counts how many times `check()`
    is called.

3.  You pass it by mutable reference to your memtest function: `test(memory, &mut iteration_counter);`

4.  When your memtest returns, if it passed then you call `iteration_counter.assert_matches()`. `IterationCounter`
    simply compares `expected_iter()` from `init()` to the number of calls to `check()` and panics if they don't match.

This should help you ensure all your counting math is right!

Also -- With a lot of the new tests, I noticed that there is a last pass that writes to memory even though it is never read again. Is there a reason for that, or are you simply matching the behavior of memtest-x86?

BrianShTsoi commented:
For the last pass that writes to memory, I am matching the behavior of memtest86+. In their [documentation](https://github.com/memtest86plus/memtest86plus?tab=readme-ov-file#memory-testing-philosophy), they talked about how this helps emulate an "ideal" memory test of "write to one memory cell, then write to all its neighbours and finally check if that one cell changed".

The tests we have might not run on the entire memory of the machine, so technically that last pass' writing to memory might not be as effective as memtest86+, but I think it is still possible that one address may be affected by writes to some neighbouring address if the hardware is faulty, so it's worth trying that.

BrianShTsoi committed:
Refactor memtest for best practices

BrianShTsoi commented:
Addressed feedback. Also rebased to newest release version.

This PR will have to be v0.2.0 as we are now using `Result<MemtestOutcome, MemtestError<TimeoutError>>` instead of `Result<MemtestOutcome, MemtestError>`

BrianShTsoi committed:
Remove unused warning on Linux

BrianShTsoi committed:
Implement unit test for counting test iterations

BrianShTsoi committed:
Correct iteration for test_mov_inv_random

BrianShTsoi commented:
Also added unit test for counting iterations.
Turns out there is indeed a memtest which I mistakenly doubled the expected iterations.

BrianShTsoi commented on src/memtest.rs:
```diff
@@ -545,3 +907,77 @@ where
+
+    // This is just a place holder, it will never be used.
+    #[derive(Debug)]
+    pub struct IterationError;
```
```suggestion
    struct IterationError;
```
Mistake when copy-pastaing

marti4d commented on src/lib.rs:
```diff
@@ -177,24 +179,10 @@ impl MemtestRunner {
-                MemtestKind::Checkerboard => memtest::test_checkerboard,
-                MemtestKind::BlockSeq => memtest::test_block_seq,
-            };
+            let test = test_kind.to_fn();
```
Hmm... I'm finding that this is getting a bit awkward now. Perhaps it would be better, instead, to move this `match()` logic to a block like this in `memtest.rs`:

```Rust
impl MemtestKind {
    pub fn run<O: TestObserver>(&self, memory: &mut usize, observer: O) -> MemtestResult<T> {
        match self {
            // ...
        }
    }
}
```

And then just have the code in `lib.rs` call it through `test_kind.run(...)`.

marti4d commented on src/memtest.rs:
```diff
@@ -545,3 +907,77 @@ where
+        completed_iter: u64,
+    }
+
+    // This is just a place holder, it will never be used.
```
Ah... Luckily, there actually exists a specific Rust type in `std` for this exact scenario: [`std::convert::Infallible`](https://doc.rust-lang.org/std/convert/enum.Infallible.html). With it, you can write `fn foo() -> Result<String, Infallible>` and Rust understands that the `Err<E>` variant will never be constructed.

In this case, you would use it like `type Error = Infallible;`.

marti4d commented on src/memtest.rs:
```diff
@@ -545,3 +907,77 @@ where
+        let mut memory = vec![0; 512];
+        for test_kind in MemtestKind::ALL {
+            let mut counter = IterationCounter::new();
+            if !matches!(
```
nit: Use `assert!(cond, "message")` instead of `if cond { panic!("message") }`

marti4d reviewed (approved):
As always, sorry about the delay. Looks great! I have a few quick cleanup suggestions, but as far as I'm concerned this is ready to land any time.

BrianShTsoi committed:
Refactor MemtestKind, MemtestFailure and unit tests for best practices

BrianShTsoi committed:
Reformat MemtestFailure Debug output

BrianShTsoi committed:
Add Block Move Test

BrianShTsoi commented on src/memtest.rs:
```diff
@@ -407,33 +416,498 @@ pub fn test_checkerboard(
+/// memtest86plus/memtest86plus?tab=readme-ov-file#memtest86-test-algorithms)
+///
+/// The test aims to stress test the memory by moving blocks of memory, such as with the `movs` instruction.
+/// It first initializes the memory with an irregular shifting pattern. Then it performs 3 memory block moves.
```
For the lack of a better phrase, I called the initial memory pattern "an irregular shifting pattern". Memtest86+ called it "shifting patterns that are inverted every 8 bytes", but that didn't make much sense to me. For reference here is the code snippet that intializes the pattern in Memtest86+, found [here](https://github.com/memtest86plus/memtest86plus/blob/main/tests/block_move.c#L62),

```
            testword_t pattern1 = 1;
            do {
                testword_t pattern2 = ~pattern1;
                write_word(p + 0,  pattern1);
                write_word(p + 1,  pattern1);
                write_word(p + 2,  pattern1);
                write_word(p + 3,  pattern1);
                write_word(p + 4,  pattern2);
                write_word(p + 5,  pattern2);
                write_word(p + 6,  pattern1);
                write_word(p + 7,  pattern1);
                write_word(p + 8,  pattern1);
                write_word(p + 9,  pattern1);
                write_word(p + 10, pattern2);
                write_word(p + 11, pattern2);
                write_word(p + 12, pattern1);
                write_word(p + 13, pattern1);
                write_word(p + 14, pattern2);
                write_word(p + 15, pattern2);
                pattern1 = pattern1 << 1 | pattern1 >> (TESTWORD_WIDTH - 1);  // rotate left
            } while (p <= (pe - 16) && (p += 16)); // test before increment in case pointer overflows
```

marti4d replied:
I agree... Your wording is better (theirs seems actually misleading), and your implementation is very nice (and seems to accurately do the same thing).

BrianShTsoi commented on src/memtest.rs:
```diff
@@ -407,33 +416,498 @@ pub fn test_checkerboard(
+
+    // TODO:
+    // This expected_iter does not account for the copy slice part, as `Observer::check()` is not
+    // called in there
+    let expected_iter = u64::try_from(memory.len())
+        .ok()
+        .and_then(|count| count.checked_mul(2))
+        .context("Total number of iterations overflowed")?;
+    observer.init(expected_iter);
+
+    let val_to_write = |pattern: usize, idx| match idx {
+        4 | 5 | 10 | 11 | 14 | 15 => !pattern,
+        _ => pattern,
+    };
+
+    // Set up initial pattern in memory
+    let mut pattern = 1;
+    for chunk in memory.chunks_exact_mut(CHUNK_SIZE) {
+        for (i, mem_ref) in chunk.iter_mut().enumerate() {
+            observer.check().map_err(MemtestError::Observer)?;
+
+            let val = val_to_write(pattern, i);
+            write_volatile_safe(mem_ref, val);
+        }
+        pattern = pattern.rotate_left(1);
+    }
+
+    // Move blocks of memory around
+    let (first_half, second_half) = split_slice_in_half(memory)?;
+    let half_len = first_half.len();
+    volatile_copy_slice(second_half, first_half);
+    volatile_copy_slice(
+        &mut first_half[OFFSET..],
+        &second_half[..(half_len - OFFSET)],
+    );
+    volatile_copy_slice(
+        &mut first_half[..OFFSET],
+        &second_half[(half_len - OFFSET)..],
+    );
+
```
I am not too sure how to approach the issue of `Observer`/`TimeoutChecker`. 

My understanding from Memtest86+ is that the original Block Move Test specifically uses `movs` to perform the block moves. I think the main goal is to stress test the memory by having more frequent memory operations than the other memory tests (although one can also make the case that this test stress tests the memory by performing many memory copies (ie. a read and then immediately a write), rather than many reads, then many writes, which is what the other tests do) 

With that context in mind, it seems that if I try to incorporate `Observer` into the block moves (ie `volatile_copy_slice`), it would prevent the compiler from optimizing it to anywhere close to something like `movs`, and this would basically defeat the point of the test. On the other hand, not incorporating it defeats the point of `Observer` in this test.

marti4d replied:
It's not perfect, but I think it's okay that we leave it out for this test. If we want, in the future, we might try to collect some stats on how long it takes this test to run and just wholesale decide "run it or not" depending on how much time we have left.

marti4d commented on src/memtest.rs:
```diff
@@ -667,6 +669,116 @@ fn mov_inv_walking_pattern<O: TestObserver>(
+}
+
+// TODO: In Memtest86+, block move is achieved with the `movs` assembly instruction
+fn volatile_copy_slice<T: Copy>(dst: &mut [T], src: &[T]) {
```
I like that you did it this way so you can eventually play around with speeding up the test. For example, it might be faster for you to use an instruction like `vmovdqa` (Which Rust provides in Stable [here](https://doc.rust-lang.org/core/arch/x86/fn._mm256_load_si256.html)).

Gian-Carlo might be able to help a lot with this. Because of all his experience working on ML with his Chess/Go engines, I'd call him an expert with SIMD instructions.

marti4d reviewed (approved):
This is great, Brian! I have nothing to add. I think you made the right choice to leave the observer checking out of your loop. We're going to have to make a bit of an exception with this test -- There's no way it will ever be acceptable to interrupt a massive memory transfer instruction like that.

In the future, we may be able to add some higher-level smarts to make the decision to go ahead or not before we ever start the transfer, but that would be another project for another day.

BrianShTsoi committed:
Add note on Observer in comments for Block Move Test

BrianShTsoi committed:
Update version number to v0.2.0
